Enable generic operation events using filter_operations.rb

sugr4 opened the pull request:
New input plugin for checking dsc install and configuration
DSC installation is checked once a day.
DSC PerformRequiredConfigurationChecks.py is run every 30 mins, two successive failures sends an operation event.
@Microsoft/omsagent-devs

robbiezhang commented on source/code/plugins/operation_lib.rb:
```diff
@@ -39,9 +39,27 @@ def filter(record, time)
-      data_item = filter(record, time)
+    def filter_and_wrap(tag, record, time)
+      case tag
+      when "oms.operation.buffer"
```
You should not determine the operation by checking the tag name. It will block the plugin from reuse.

robbiezhang commented on source/code/plugins/operation_lib.rb:
```diff
@@ -39,9 +39,27 @@ def filter(record, time)
+        return {
+          "Timestamp"=>OMS::Common.format_time(time),
+          "OperationStatus"=>"Warning",
+          "Computer"=>`hostname`.strip,
```
use OMS::Common.get_hostname

robbiezhang commented on source/code/plugins/in_dsc_checks.rb:
```diff
@@ -0,0 +1,114 @@
+require 'fluent/config/error'
+
+module Fluent
+  class DscInput < Input
```
DscMonitoringInput

robbiezhang commented on source/code/plugins/in_dsc_checks.rb:
```diff
@@ -0,0 +1,114 @@
+
+module Fluent
+  class DscInput < Input
+    Fluent::Plugin.register_input('dsc_checks', self)
```
dsc_monitor

robbiezhang commented on source/code/plugins/in_dsc_checks.rb:
```diff
@@ -0,0 +1,114 @@
+
+    def get_dsc_status
+      begin
+        dsc_status = %x(/opt/microsoft/omsconfig/Scripts/PerformRequiredConfigurationChecks.py)
```
Run TestDscConfiguration.py to get the status.

robbiezhang commented on source/code/plugins/in_dsc_checks.rb:
```diff
@@ -0,0 +1,114 @@
+      rescue => error
+        OMS::Log.error_once("Unable to run PerformRequiredConfigurationChecks.py for dsc : #{error}")
+      end
+      if dsc_status.include?("ReturnValue=0")
```
This is a sample error state:
instance of TestConfiguration
{
    ReturnValue=0
    InDesiredState=false
    ResourceId={}
}

robbiezhang commented:
not sign off, as the dsc test script is wrong.

sugr4 commented:
made changes based on comments

robbiezhang commented on source/code/plugins/in_dsc_monitor.rb:
```diff
@@ -0,0 +1,114 @@
+
+  end
+
+  class DSCcache
```
it's still an in-memory cache. we need to use file to persist the last status because the DSC will restart the OMS agent when it's not in desired state.

robbiezhang commented on source/code/plugins/operation_lib.rb:
```diff
@@ -39,9 +39,28 @@ def filter(record, time)
-    def filter_and_wrap(record, time)
-      data_item = filter(record, time)
+    def filter_and_wrap(tag, record, time)
+      tag_type = tag.sub(/^oms.operation./,'')
```
can we use the last part of the tag as the tag_type? don't assume that it starts with oms.operation.

varunkumta commented on source/code/plugins/operation_lib.rb:
```diff
@@ -39,9 +39,28 @@ def filter(record, time)
+        dataitem["Timestamp"] = OMS::Common.format_time(time)
+        dataitem["OperationStatus"] = "Warning"
+        dataitem["Computer"] = OMS::Common.get_hostname or "Unknown host"
+        dataitem["Detail"] = record["message"]
```
also populate Category and Solution fields .. @agup006 can you specify them?

varunkumta replied:
and they should probably come from the record emitted by the input plugin itself

agup006 replied:
@sugr4 @varunkumta 

- Category = OMS Agent for Linux issue
- Solution = Log Management

sugr4 commented:
Used YAML::Store's persistent caching to store dsc status
Changed tag parsing for operation_lib
Added Category and Solution fields for dsc operations

robbiezhang commented on source/code/plugins/in_dsc_monitor.rb:
```diff
@@ -0,0 +1,112 @@
+        install_status = check_install
+
+        # store dsc installation status in the cache
+        @dsc_cache.transaction {
```
you don't need to cache the install status. you need to get it every time anyway.

sugr4 commented:
Added emit message for when dsc is not installed

sugr4 commented:
Changed tag for monitor_agent plugin to be oms.health.** and kept oms.operation.** tag for other operations to have clearer separation between monitor.conf and operation.conf

robbiezhang commented:
Signoff
